### PR TITLE
Complete MOD to Ext conversion

### PR DESCRIPTION
**Pull request by VSEphpbb**
This should complete the MOD to ext convert :+1: 

It's working in 3.1 now. Only outstanding issues to solve:
~~- [ ] The hard-coded constants for FORUM_ID and USER_ID from the config.php~~ (in another PR)
- [x] The Idea BBCodes (refactored to use the `url` bbcode)

@marc1706 
@callumacrae 
@michaelcullum 
@rxu

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,27 +1,28 @@
...
+    "description": "phpBB Ideas extension",
+    "homepage": "https://www.phpbb.com",
+    "version": "2.0.0-dev",
+    "license": "GPL-2.0",
+    "authors": [
+        {
+            "name": "Callum Macrae",
+            "email": "callum@phpbb.com",
+            "homepage": "http://macr.ae/",
+            "role": "Head Fixer of Things"
```
Comment (callumacrae) [inline composer.json]: I should probably be removed from this :joy:

**Comment by callumacrae** [comment]
Awesome :+1:

File: composer.json (hunk 1 above)
Comment (VSEphpbb) [inline composer.json]: 9%% is still your code... Well, hmm, considering the bulk of it is copied from viewtopic... :smile: 

But it's mostly just conversion not rewrite here.

**Comment by VSEphpbb** [comment]
I don't see why we can't just use the core's `url` bbcode to achieve the same effect as the custom ideas bbcodes now, to create links in the topics to the idea.

I also think ti would be good to add the new front end changes PR #1 into this as well so it looks more up to date too.

**Comment by michaelcullum** [comment]
Wow. Thanks for this. I'll have a look over in a wee bit.

**Comment by VSEphpbb** [comment]
@michaelcullum Cool. I think I'm done pushing changes to this PR, though I will work on the front end improvements by Christian for another PR after this one.

**Comment by rxu** [comment]
Looks good in common.
1) It doesn't seem to have a link to the ```/ideas``` page anywhere, so it would be nice to add.
2) When there're no ideas, the page looks like that:
ideas
It's probably worth adding something like ```No ideas``` here with getting rid of the header bars with column titles, as well as ```View all``` link in case of no ideas.

**Comment by VSEphpbb** [comment]
@rxu 
1) Thats intended for now. Only because the original MOD did not have a link to ideas either. I'm not sure et how phpBB handled this in the past or plans to for the future. It may be a link in phpBB.com's main menu system for example, rather than something in the forum itself.
2) That's already been addressed and will be added here soon with some other visual updates.

**Comment by VSEphpbb** [comment]
Updated: I went ahead and added the new improved voting buttons and some other visual updates to this PR after all.

File: README.md [hunk 2]
```diff
@@ -4,31 +4,25 @@ phpBB Ideas is an ideas centre for phpBB. It is based on [WordPress ideas](http:
...
-
 ## Contributing
 
 Please fork this repository and submit a pull request to contribute to phpBB Ideas
 
 ## Bug Reporting & Support
 
-You can report bugs or suggest features in the [issue tracker](https://www.phpbb.com/bugs/website/).
-Support is not available for phpBB Ideas however you may email `website [at] phpbb [dot] com` where you may recieve some limited support should the problem be with phpBB Ideas.
+You can report bugs or suggest features in the [issue tracker](https://github.com/phpbb/phpbb-ideas/issues).
```
Comment (michaelcullum) [inline README.md]: Should be the main website tracker on JIRA still

File: composer.json (hunk 1 above)
Comment (michaelcullum) [inline composer.json]: Let's change the role here, add 'phpBB Website Team' and 'phpBB Extensions Team' and VSE add yourself.

File: config/services.yml [hunk 3]
```diff
@@ -11,10 +11,46 @@ services:
...
+
+    phpbb.ideas.idea_controller:
+        class: phpbb\ideas\controller\idea_controller
+        arguments:
+            - @auth
+            - @cache
+            - @config
+            - @content.visibility
+            - @dbal.conn
+            - @controller.helper
```
Comment (michaelcullum) [inline config/services.yml]: ordering

File: config/services.yml [hunk 4]
```diff
@@ -11,10 +11,46 @@ services:
...
+            - @phpbb.ideas.Ideas
+            - @request
+            - %core.root_path%
+            - %core.php_ext%
+
     phpbb.ideas.Ideas:
         class: phpbb\ideas\factory\Ideas
         arguments:
             - @dbal.conn
+            - @controller.helper
```
Comment (michaelcullum) [inline config/services.yml]: and here

File: controller/base.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\ideas\controller;
+
+// @todo: refactor out
```
Comment (michaelcullum) [inline controller/base.php]: Ermmm

**Comment by marc1706** [comment]
I was going to take a look at this but this is far as I get while testing this: Class phpbb\ideas\factory\LinkHelper does not exist

File: controller/index_controller.php [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+				'POINTS'    => $row['idea_votes_up'] - $row['idea_votes_down'],
+			));
+		}
+
+		$rows = $this->ideas->get_ideas(10, 'top', 'DESC');
+		foreach ($rows as $row)
+		{
+			$this->template->assign_block_vars('top_ideas', array(
+				'ID'		=> $row['idea_id'],
+				'LINK'		=> $this->link_helper->get_idea_link($row['idea_id']),
```
Comment (michaelcullum) [inline controller/index_controller.php]: This should be done in the template

File: controller/index_controller.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+				'TITLE'		=> $row['idea_title'],
+				'AUTHOR'	=> $this->link_helper->get_user_link($row['idea_author']),
```
Comment (michaelcullum) [inline controller/index_controller.php]: And this

**Comment by marc1706** [comment]
Seems to work well apart from the above mentioned issue caused by the class names :+1:

**Comment by michaelcullum** [comment]
```
<MichaelC> the diff is kinda hard to read
<VSE> I know. This was not a rewrite… it was a refactor from MOD to ext is all, with any needed modernizations along the way
<MichaelC> yeah
<MichaelC> but i've got no idea what has changed
<MichaelC> which makes reviewing really difficult
<VSE> I know, sorry…
<MichaelC> no worries
<MichaelC> i'm just going to leave all the files that are new/deleted
<MichaelC> and i'll just read those marked as modified
<MichaelC> otherwise i'm going to accidentally review all callum's stuff
<MichaelC> which will take forever
<VSE> lol
```

File: factory/ideas.php [hunk 8]
```diff
@@ -1,31 +1,73 @@
 <?php
 /**
  *
- * @package phpBB3 Ideas
+ * This file is part of the phpBB Forum Software package.
```
Comment (marc1706) [inline factory/ideas.php]: Is it?

File: controller/base.php (hunk 5 above)
Comment (VSEphpbb) [inline controller/base.php]: Will be refactored out in a separate PR after this one is merged, adding a new ACP module to choose a forum and user.

**Comment by VSEphpbb** [comment]
Updated. All comments addressed. Note the item about refactoring the constants will be a new PR after this one, adding an ACP module to set them.

File: composer.json (hunk 1 above)
Comment (callumacrae) [inline composer.json]: BORING

Comment (michaelcullum) [inline composer.json]: Some of us have other ways to amuse ourselves than roles in a composer.json file.
